### PR TITLE
Issue #3843: Add check for missing classes in Cobertura coverage report

### PR DESCRIPTION
**Pull request by MEZk**
Issue #3843
Added check for missing classes in Cobertura coverage report

### CODE REVIEW

**Comment by rnveach** [comment]
https://travis-ci.org/checkstyle/checkstyle/jobs/204567133#L1083-L1090
Aren't we missing `RequireThisCheck` being reported because of #3848 ?

**Comment by MEZk** [comment]
@rnveach 
Yes, problem in regexp.

Do you know why Travis CI does not fail the build?
https://github.com/checkstyle/checkstyle/pull/3853/files#diff-1625ee7bf90900416b7d197caa88c370R185
https://travis-ci.org/checkstyle/checkstyle/jobs/204569528#L1083

File: .ci/travis/travis.sh [hunk 1]
```diff
@@ -171,6 +171,21 @@ no-exception-test-alot-of-project1)
...
+  grep "nowrap" target/site/cobertura/frame-sourcefiles.html | grep -Po "[A-Za-z.]+\.html" | sed 's/\.html//g' | sed 's/\./\//g' | sort | uniq > cobertura_classes.log
+  find target/classes -type f -name "*.class" | grep -vE ".*\\$.*" | sed 's/target\/classes\///g' | sed "s/.class//" | sort | uniq > target_classes.log
+  sed -n '/<instrumentation/,/<\/instrumentation/p' pom.xml | grep "<exclude>.*</exclude>" | grep -Po "[A-Za-z*/]+\*?.class" | sed 's/\.class//g' | sed 's/\*//g' | sort | uniq > cobertura_excluded_classes.log
+  grep -Fxv -f cobertura_classes.log target_classes.log > missed_classes_with_excludes.log
+  grep -Fv -f cobertura_excluded_classes.log missed_classes_with_excludes.log > missed_classes_without_excludes.log
+
+  if [[ -s missed_classes_without_excludes.log ]] ; then
+	echo "Classes which are missed in Cobertura coverage report:"
+	cat missed_classes_without_excludes.log
+	exit 1;
```
Comment (rnveach) [inline .ci/travis/travis.sh]: should the semi-colon be there? Line 191 below doesn't have it.

**Comment by MEZk** [comment]
frame-sourcefiles.html cannot be use to grab all the classes which are instrumented by cobertura. The file contains information about RquireThisCheck but the check is not included in the report.

I guess that coverage.xml should be a valid file.

File: .ci/travis/travis.sh (hunk 1 above)
Comment (MEZk) [inline .ci/travis/travis.sh]: Removed semi-colon and faced the same problem.

**Comment by romani** [comment]
Teamcity complain should be fixed

**Comment by rnveach** [comment]
@romani We don't know what the issue is that it isn't failing. Travis is outputing failing scenario, but it doesn't exit with 0.

There is still a separate issue that `RequireThisCheck` isn't part of the output.

**Comment by MEZk** [comment]
I fixed problem with RequireThisCheck, please take a look at the report on Travis

**Comment by romani** [comment]
interesting case, looks like "set -e" is not set for Travis generated scripts any more,  to let fail all on any failure inside.
@MEZk , please check exit code right after "eval $CMD" (".travis.yml"). 
We have two options:
- set "set -e" in travis 
- manage exit code ourself to deliver it to Travis.

File: .ci/travis/travis.sh [hunk 2]
```diff
@@ -171,6 +171,21 @@ no-exception-test-alot-of-project1)
   groovy ./launch.groovy --listOfProjects projects-for-travis.properties --config checks-nonjavadoc-error.xml
   ;;
 
+cobertura-check)
+  mvn clean compile cobertura:check cobertura:cobertura
+  grep -Po "<class name=\".+\"" target/site/cobertura/coverage.xml | grep -Po "com.puppycrawl.tools.checkstyle\.?.*\" filename" | sed 's/\" filename//g' | sed 's/\./\//g' | sort | uniq > cobertura_classes.log
```
Comment (romani) [inline .ci/travis/travis.sh]: It is ok, but ...
I do not recommend to use grep for xml parsing, it is not stable solution
Consider to use xmlstarlet, youcan find example of usages in this repo and sevnu.

Comment (MEZk) [inline .ci/travis/travis.sh]: Done

**Comment by MEZk** [comment]
@romani 
> @MEZk , please check exit code right after "eval $CMD" (".travis.yml").

Done. 
https://github.com/checkstyle/checkstyle/pull/3853/commits/5677bf6b15c40c6ba3c2acbac1acb9d44eefc8a3#diff-1625ee7bf90900416b7d197caa88c370R185
https://github.com/checkstyle/checkstyle/pull/3853/commits/5677bf6b15c40c6ba3c2acbac1acb9d44eefc8a3#diff-354f30a63fb0907d4ad57269548329e3R168

It fails the build but does not print any information.

**Comment by romani** [comment]
I noticed the same problem of ignorance of exit code at https://travis-ci.org/checkstyle/checkstyle/jobs/204781086#L301 , this Build should be red.

**Comment by MEZk** [comment]
@romani 
Why we set ```set -e``` in travis.sh rather than in travis.yml?

**Comment by romani** [comment]
it was not a problem before, I do not know what was changed on Travis recently.

**Comment by MEZk** [comment]
@romani 
> Please add command "set -e" before "- SKIP_FILES=".github|appvey........"

Fails the build but does not print info into log.

**Comment by romani** [comment]
I do not see logs at all https://travis-ci.org/checkstyle/checkstyle/jobs/204982940

**Comment by romani** [comment]
@MEZk , I fixed Travis , please rebase

**Comment by MEZk** [comment]
@romani 

> @MEZk , I fixed Travis , please rebase

Done.
Still Travis does not execute the [last grep command](https://github.com/checkstyle/checkstyle/pull/3853/commits/9105390111c7b4f96cf3e0b318de537d80b8cb71#diff-1625ee7bf90900416b7d197caa88c370R189) and [if-else commands](https://github.com/checkstyle/checkstyle/pull/3853/commits/9105390111c7b4f96cf3e0b318de537d80b8cb71#diff-1625ee7bf90900416b7d197caa88c370R193)

See logs:
https://travis-ci.org/checkstyle/checkstyle/jobs/205224483#L1090

**Comment by romani** [comment]
>  There were not any problems when I used grep instead of xmlstarlet.
Can you check the script on your machine?

Travis is a priority and it is etalon. We need to use ways that works for Travis and locals.
I will take a look at srcipt, please return to grep for now.

**Comment by MEZk** [comment]
@romani 
>  or just add one more commands to correct output of old xmlstarlet (and explain it in comment). I like second approach more.

Done. The problem was fixed. The build will be fixed after https://github.com/checkstyle/checkstyle/pull/3849 .

**Comment by romani** [comment]
#3849 is merged.

@MEZk , please do rebase to make PR green.

**Comment by MEZk** [comment]
@romani 
> @MEZk , please do rebase to make PR green.

Done. And now Travis does not execute the following command:
https://github.com/checkstyle/checkstyle/pull/3853/commits/4bf61e02c55bc77cbb6aec899881d2bed2453601#diff-1625ee7bf90900416b7d197caa88c370R193

However, all files required were created:
1) https://travis-ci.org/checkstyle/checkstyle/jobs/205407262#L1848
2) https://travis-ci.org/checkstyle/checkstyle/jobs/205407262#L1869

I do not know what causes the problem.

**Comment by romani** [comment]
@MEZk , it is executed and probably return non zero exit code, that is whole Travis execution is stopped, so you do not have any logs.
you need to do "grep ..... | cat" to eliminate exit code from grep. If grep is failing to return value it return non zero exit code.

**Comment by MEZk** [comment]
@romani 
I added ```cat``` at the end of the grep pipeline
https://github.com/checkstyle/checkstyle/pull/3853/commits/d12fa314a25e7b5624d18e61ba79ee03355d0774#diff-1625ee7bf90900416b7d197caa88c370R183

and it fixed the problem
https://travis-ci.org/checkstyle/checkstyle/jobs/205470504#L1093

Still do not understand why. Should I add ```| cat``` at the end of all grep commands of the script?

**Comment by romani** [comment]
> Should I add | cat at the end of all grep commands in the script?

"... | cat" is useful only when it is expected that content could be missed. So not all grep commands should have piped "cat".
in case of "grep -Fxvf cobertura_classes.log target_classes.log > missed_classes_with_excludes.log"
if output is nothing - it mean that smth is broken before and it is not expected.

> grep -Fvf cobertura_excluded_classes.log missed_classes_with_excludes.log > missed_classes_without_excludes.log | cat

please change to "grep ..... | cat > output-file"

**Comment by MEZk** [comment]
@romani 
> please change to "grep ..... | cat > output-file"

Done. See https://github.com/checkstyle/checkstyle/pull/3853/commits/77dda2b7670536e1408fde1ded01ad6d9caad63b#diff-1625ee7bf90900416b7d197caa88c370R183-185

output file is empty
https://travis-ci.org/checkstyle/checkstyle/jobs/205510567#L1090

**Pull request merged**